jabbercat/models.py

```diff
@@ -790,43 +789,44 @@ def _tags_set_changed(self):
         self.invalidateFilter()
 
     @property
-    def filter_by_text(self):
-        return self._filter_by_text
+    def filter_text_input(self):
+        return self._filter_text_input
 
-    @filter_by_text.setter
-    def filter_by_text(self, value: str):
-        self._filter_by_text = self._normalize_for_find(value)
+    @filter_text_input.setter
+    def filter_text_input(self, value: str):
+        self._filter_text_input = self._normalize_for_comparison(value)
         self.invalidateFilter()
 
-    @filter_by_text.deleter
-    def filter_by_text(self):
-        self._filter_by_text = None
-
-    def _contains_normalized(self, normed: str, other: str):
-        return normed in self._normalize_for_find(other)
+    @filter_text_input.deleter
+    def filter_text_input(self):
+        self._filter_text_input = None
 
     def filterAcceptsRow(self,
                          source_row: int,
                          source_parent: Qt.QModelIndex):
         source = self.sourceModel()
-        item = source.data(
+        item: AbstractRosterItem = source.data(
             source.index(source_row, 0, source_parent), ROLE_OBJECT
         )
 
         if isinstance(item, jclib.roster.SubscriptionRequestItem):
             # filter inbound subscription requests
             return False
 
-        if self._filter_by_text is not None:
-            if (not self._contains_normalized(self._filter_by_text,
-                                              str(item.address)) and
-                    not self._contains_normalized(self._filter_by_text,
-                                                  item.label)):
-                return False
+        if self._filter_text_input is not None:
+            query_args = [[str(item.address)], [item.label], item.tags]
+            normalized = [self._normalize_for_comparison(s)
+                          for s in chain.from_iterable(query_args)]
+            text_input = self._normalize_for_comparison(self._filter_text_input)
+            return any(text_input in n for n in normalized)
 
-        filter_tags = self._tags_filter_set.checked
-        if set(item.tags) & filter_tags != filter_tags:
-            return False
+        # Todo: popose to remove the "checked tag" logic when clicking on a tag,
+        #  instead it may be better to populate the search bar with "tag:'the_tag'"
+        #  when clicking on a tag (-> a single place for filtering, more discoverable).
```

Comment:
I am not sure about this. This might make sense for an advanced UI, but I’m not sure if it makes sense for normal users. Especially in the context of i18n, doing "magic" things with parts of text will become weird quickly.

Clicking on tags to filter for them has a precedent in the web though.

I agree that we need a way to do this filtering using text input, even if only for accessibility. My suggestion for that would be the magic drop down menu/completion thing I mentioned elsewhere.

Reply:
The ideal solution would be to have the tags (styled as tags) appear in the filter bar and allow completion of filter text to tags (which would work nicely with the drop-down-menu). This would solve the quoting problem (how to mark a tag search vs. search in other text, how to escape that marking), be discoverable and nice (I don't like that we currently do not show which tag filters are active). However, it might be quite challenging to implement the rendering of the tags in the search bar.

Reply:
@sebastianriese Indeed, that would be the most amazing solution.

As you correctly identified, this is extremely tricky. I mean, it is *easy* to have the tags on the left of the input, we had that at one point. But having them inside the input is much more tricky and interaction with the text cursor becomes interesting to say the least.

I’d very much like to not have to re-write the text layouting of QLineEdit (even though that’s at least less scary than re-writing the text layouting of QTextEdit).

Reply:
Hmm perhaps having them to the left is fine. If they are added by clicking or by completion I think it would be acceptable to have them jump to a canonical position (and perhaps order). Of course this may be not ideal for the accessibility (since the tags would no longer be simple text).

Reply:
There are several issues:

* Filtering for many/long tags makes the input shrink (because the tags don’t scroll like normal text in the input does)
* Editing of the selected tags is not straightforward using the keyboard only; they are not part of the text input, thus you can’t navigate to them using arrow keys. Deletion has similar issues.

The second point could be fixed by hooking into keyPressEvent and give focus to the rightmost tag widget when the cursor is leftmost in the text input and Qt.Qt.Left is pressed. Likewise, pressing Qt.Qt.Right on the rightmost tag widget should hand focus to the text input (and position the cursor to the left). Transferring focus between the tag widgets would work with Left/Right too.

The tag widgets could erase themselves (and remove the tag from the filter list) on backspace/delete key, and of course also show their focus state visually. This could actually provide sufficient accessibility.

The first problem isn’t as easy to solve. It may be possible to apply some trickery with layouts, scaling and a nice scrolling solution for the tag widgets.

I haven’t had the idea of passing focus with Left/Right and using Backspace to delete tags, I think this could actually become usable that way. Maybe we could go as far as to use a QColumnView for the tags (which would give us free scrolling and also easy handling of the tags themselves).

I think we should at least explore this possibility further!

Reply:
I would prefer to not have the text parsing in it at all. In fact, I think this is a very bad idea. The only reason for doing this text-based is if we want users to manipulate these strings. I don’t want users to manipulate these strings.

I want to have a proper UI for that. If you need to filter by tags, this is possible today by either clicking on them or by using the right-click menu on the roster.

For now, it is good if we *also* search for matches inside tags (which is what I thought this PR was about).

Reply:
Agreed, the (maybe too subtle) question here was "what actually is the scope of this PR", or, "is this PR functionally satisfactory or does it need more work"

Reply:
I would be satisfied with a PR which:

1. keeps the current click filtering intact
2. adds code which makes the text input *also* search in tags (in addition to searching in jids and labels)

Reply:
OK